[projects] Project examples and documentation (#5407)
ray-project/ray: doc/source/projects.rst

Ray Projects (Experimental)
===========================

Ray projects make it easy to package a Ray application so it can be
rerun later in the same environment. They allow sharing and reuse

> existing code reliably.

Reviewer: They allow for the sharing and reuse of existing code reliably.

Author: fixed

Quick start (CLI)
-----------------

.. code-block:: bash

    # Creates a project in the current directory. It will create a
    # project.yaml defining the code and environment and a cluster.yaml
    # describing the cluster configuration.
    $ ray project create <project-name>

>     # Create a new session from the given project with an optional name.

Reviewer: What happens if you don't give it a name? autogen'd?

Author: At the moment it will be autogenerated as the project name, once we support multiple concurrent sessions we should have a smarter way to autogenerate it.

>     # This will launch a cluster and run the appropriate command.

Reviewer: ```suggestion
    # Launch a cluster and run the appropriate command.
```

Stay consistent with above.

    $ ray session start [--name="default-session"]

>     # Open a console for the given session. If the session is stopped

Reviewer: ```suggestion
    # Open a console for the given session. If the session is stopped,
```

    # this auto-starts the session if --start is specified.
    $ ray session attach [<default-session>] [--start]

>     # Stop the given session, and all of its worker nodes. The nodes/clusters

Reviewer: ```suggestion
    # Stop the given session and all of its worker nodes. The nodes/clusters
```

>     # are not actually terminated unless --terminate is given or unless we

Reviewer: ```suggestion
    # are not actually terminated unless --terminate is given or we
```

    # configure the cluster to autoterminate after a period of inactivity.
    $ ray session stop [<default-session>] [--terminate]

Examples
--------
- `Open Tachotron <https://github.com/ray-project/ray/blob/master/python/ray/projects/examples/open-tacotron/.rayproject/project.yaml>`__:
  A TensorFlow implementation of Google's Tacotron speech synthesis with pre-trained model (unofficial)

> - `Mozilla DeepSpeech <https://github.com/ray-project/ray/blob/master/python/ray/projects/examples/mozilla-deepspeech/.rayproject/project.yaml>`__:

Reviewer: Would be good to have one TF and one other

  A TensorFlow implementation of Baidu's DeepSpeech architecture

Project file format (project.yaml)
----------------------------------

> A project file contains all information, including cluster configuration,

Reviewer: A bit awkward. Replacement suggestion:

"A project file contains everything required to run a project. This includes a cluster configuration, the environment and dependencies for the application, and the specific inputs used to run the project."

Author: fixed

environment and dependencies, and inputs required to run a project.

Here is an example for a minimal project format:

.. code-block:: yaml

    name: <<project name>>
    description: <<short description>>
    repo: <<code repository, e.g., github repo>>

    # Cluster to be instantiated by default when starting the project.
    cluster: <<autoscaler config file>>

    # Commands/information to build the environment, once the cluster is
    # instantiated. This can include the versions of python libraries etc.
    # It can be specified as a Python requirements.txt, a conda environment,

>     # a Docker file or a shell script to run to set up the libraries.

Reviewer: ```suggestion
    # a Dockerfile, or a shell script to run to set up the libraries.
```

Ray Projects (Experimental)
===========================

Ray projects make it easy to package a Ray application so it can be
rerun later in the same environment. They allow for the sharing and
reuse of existing code reliably.

Quick start (CLI)
-----------------

.. code-block:: bash

    # Creates a project in the current directory. It will create a

>     # project.yaml defining the code and environment and a cluster.yaml

Reviewer 2: Is it possible to infer the env based on current conda installed packages?

Author: Not yet but should be, I created an issue here: https://github.com/ray-project/ray/issues/5487

>     # project.yaml defining the code and environment and a cluster.yaml

Reviewer 3: Specify where these files will be made?

    # describing the cluster configuration.
    $ ray project create <project-name>

    # Create a new session from the given project with an optional name.
    # Launch a cluster and run the appropriate command.
    $ ray session start [--name="default-session"]

    # Open a console for the given session. If the session is stopped,
    # this auto-starts the session if --start is specified.
    $ ray session attach [<default-session>] [--start]

    # Stop the given session and all of its worker nodes. The nodes/clusters
    # are not actually terminated unless --terminate is given or we
    # configure the cluster to autoterminate after a period of inactivity.
    $ ray session stop [<default-session>] [--terminate]

Examples
--------

> - `Open Tachotron <https://github.com/ray-project/ray/blob/master/python/ray/projects/examples/open-tacotron/.rayproject/project.yaml>`__:

Reviewer 3: ```suggestion
- `Open Tacotron <https://github.com/ray-project/ray/blob/master/python/ray/projects/examples/open-tacotron/.rayproject/project.yaml>`__:
```

Ray Projects (Experimental)
===========================

Ray projects make it easy to package a Ray application so it can be
rerun later in the same environment. They allow for the sharing and
reliable reuse of existing code.

Quick start (CLI)
-----------------

.. code-block:: bash

    # Creates a project in the current directory. It will create a
    # project.yaml defining the code and environment and a cluster.yaml
    # describing the cluster configuration. Both will be created in the
    # .rayproject subdirectory of the current directory.
    $ ray project create <project-name>

    # Create a new session from the given project with an optional name.
    # Launch a cluster and run the appropriate command.

>     $ ray session start [--name="default-session"]

Reviewer 4: these optional arguments are not implemented in the CLI PR. because multiple session snapshot is not implemented. Can we remove them in doc and add these back when they are implemented? 

```diff
- ray session start [--name="default-session"]
+ ray session start 

- ray session attach [<default-session>] [--start]
+ ray session attach 

- ray session stop [<default-session>] [--terminate]
+ ray session stop
```
